Feature/bjs2 66307 check post

marsel-mkh opened the pull request:
Проверку выполнил а дебаге с кроной в 10 секунд

Commit:
BJS-66307-check-post task compl v1

Commit:
BJS-66307-check-post task with test

Commit:
BJS-66307-check-post task with test

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -206,4 +218,84 @@ public Post getPostEntity(long postId) {
+     */
+    @Async
+    @Transactional
+    public CompletableFuture<Void> checkForProfanity(List<Post> posts) {
```
Если я не ошибаюсь, то вроде как, если ты помечаешь метод аннотацией `@Async`, то использовать его в этом же классе нельзя, иначе метод не будет вызываться асинхронно, так что стоит вынести данный метод в отдельный класс

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -206,4 +218,84 @@ public Post getPostEntity(long postId) {
+            return;
+        }
+
+        List<List<Post>> batches = partition(unverifiedPosts, batchSize);
```
можно просто использовать ListUtils.partition()

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -206,4 +218,84 @@ public Post getPostEntity(long postId) {
+
+        List<List<Post>> batches = partition(unverifiedPosts, batchSize);
+
+        List<CompletableFuture<Void>> futures = batches.stream()
```
может назвать, например, moderationTasks?

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -206,4 +218,84 @@ public Post getPostEntity(long postId) {
+    public CompletableFuture<Void> checkForProfanity(List<Post> posts) {
+        try {
+            Set<String> profanity = moderationDictionary.getProfanityWord();
+            LocalDateTime now = LocalDateTime.now();
```
Может получение словаря как-то вынести? думаю сейчас это немного неэффективно

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -206,4 +218,84 @@ public Post getPostEntity(long postId) {
+    @Transactional
+    public CompletableFuture<Void> checkForProfanity(List<Post> posts) {
+        try {
+            Set<String> profanity = moderationDictionary.getProfanityWord();
```
Может получение словаря как-то вынести? думаю сейчас это немного неэффективно

gulnaz21 commented on src/test/java/faang/school/postservice/service/PostServiceTest.java:
```diff
@@ -332,33 +348,68 @@ public void givenInvalidProjectIdWhenGetProjectPublishedPostThenReturnListOfPost
     }
+
+    @Test
+    @DisplayName("Проверка модерации поста, пост промодерирован и прошел верефикацию")
```
верификация

gulnaz21 commented on src/test/java/faang/school/postservice/service/PostServiceTest.java:
```diff
@@ -332,33 +348,68 @@ public void givenInvalidProjectIdWhenGetProjectPublishedPostThenReturnListOfPost
+        ReflectionTestUtils.setField(postService, "batchSize", 4);
+
+        assertThrows(RuntimeException.class, () -> postService.moderateUnverifiedPost());
+    }
```
можно добавить тест, если список постов для модерации пустой

Commit:
BJS-66307-check-post added test
added PostModerationConfig, PostModerationService

Commit:
BJS-66307-check-post
fix: added PostModerationAsyncHandler and him test

RinatBeybutov commented on src/main/java/faang/school/postservice/dictionary/ModerationDictionary.java:
```diff
@@ -0,0 +1,79 @@
+public class ModerationDictionary {
+    private final PostModerationConfig postModerationConfig;
+
+    private final Set<String> profanityWord = new HashSet<>();
```
words тк их много

RinatBeybutov commented on src/main/java/faang/school/postservice/dictionary/ModerationDictionary.java:
```diff
@@ -0,0 +1,79 @@
+            loadDictionary();
+        } catch (IOException e) {
+            log.error("Failed to load moderation dictionary. Shutting down.", e);
+            System.exit(1);
```
давай уберем это, в будущем вызовет проблемы, если у нас есть перезапуск сервисов в кубере

RinatBeybutov commented on src/main/java/faang/school/postservice/dictionary/ModerationDictionary.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    @Cacheable(value = "profanityDictionary", sync = true)
+    public Set<String> getProfanityWord() {
+        return new HashSet<>(profanityWord);
```
здесь можно использовать Set.copyOf(..) там неизменяемая коллекция внутри выдается) для себя попробуй изменить неизменяемую коллекцию и посмотреть что будет

RinatBeybutov commented on src/main/java/faang/school/postservice/mapper/PostMapper.java:
```diff
@@ -38,14 +39,23 @@ public interface PostMapper {
     default List<Long> getCommentIds(List<Comment> comments) {
-        return comments.stream().map(Comment::getId).toList();
+        if (comments != null) {
+            return comments.stream().map(Comment::getId).toList();
```
каждая операция стрима с новой строчки

RinatBeybutov commented on src/main/java/faang/school/postservice/mapper/PostMapper.java:
```diff
@@ -38,14 +39,23 @@ public interface PostMapper {
     default List<Long> getResourcesIds(List<Resource> resources) {
-        return resources.stream().map(Resource::getId).toList();
+        if (resources != null) {
+            return resources.stream().map(Resource::getId).toList();
```
каждая операция стрима с новой строчки

RinatBeybutov commented on src/main/java/faang/school/postservice/service/PostModerationAsyncHandler.java:
```diff
@@ -0,0 +1,63 @@
+    @Async
+    public CompletableFuture<Void> checkForProfanity(@NotNull List<Post> posts) {
+        try {
+            Set<String> profanity = moderationDictionary.getProfanityWord();
```
вместо profanity давай  назовем words

RinatBeybutov commented on src/main/java/faang/school/postservice/service/PostModerationService.java:
```diff
@@ -0,0 +1,48 @@
+                    .anyMatch(content::contains);
+
+            post.setVerifiedAt(now);
+            post.setVerified(!profanityFound);
```
в отрицаниях легко запутаться, подумай, как переделать на положительное. Для этого мб нужно поменять название переменной и использовать другую операцию стрима

Commit:
BJS-66307-check-post
fix: replace anyMatch with noneMatch
replace hashset with immutable collection in getProfanityWord

RinatBeybutov commented:
нужно вмержить изменения и можно мержить

Commit:
Merge branch 'murloc-master-stream9' into feature/BJS2-66307-check-post

# Conflicts:
#	src/main/resources/application.yaml
#	src/main/resources/db/changelog/db.changelog-master.yaml

Commit:
BJS-66307-check-post
fix убрал одинаковые методы, решил конфликты

Commit:
BJS-66307-check-post
fix рефактор
